## Experimenting against Iris dataset - Modified ICQ Method

Runs the modified ICQ classifier against Iris dataset, using Stratified 10-Fold cross validation throughout many different random seeds to validate the classifier. At the end of each 10-Fold cross validation, it prints the AVG score and F1-Score for the classifier - and after executing with all different random seeds, the average score/f1-score is also printed.

In [3]:
import sys
import os
sys.path.append(os.path.abspath('../../'))
sys.path.append(os.path.abspath('../../models'))
sys.path.append(os.path.abspath('../../helpers'))

In [4]:
import numpy as np
from helpers.utils import print_metrics, executeIrisOneVsRest

In [5]:
import warnings
from sklearn.exceptions import UndefinedMetricWarning

# We're ignoring some warning from sklearn.metrics.classification_report
warnings.simplefilter(action='ignore', category=UndefinedMetricWarning)

## Normal SigmaQ - sigmaX + sigmaY + sigmaZ

### Importing function

In [4]:
from helpers.icq_executions import execute_classifier_split_input_weight_normal_sigma_q

### Sigma_Q param = [1, 1, 1]
We're first executing this test for different seeds using the same sigma Q as the original one, but with the new approach (having weights on U operator and inputs on rho env)

In [5]:
%%time
scores = []
f1scores = []
for i_random_state in range(10, 70, 5):
    curr_scores, curr_f1scores = executeIrisOneVsRest(random_seed=i_random_state,
                                                classifier_function=execute_classifier_split_input_weight_normal_sigma_q,
                                                sigma_q_weights=[1, 1, 1],
                                                max_iter=2000,
                                                print_each_fold_metric=False,
                                                print_avg_metric=True)
    scores.append(np.mean(curr_scores))
    f1scores.append(np.mean(curr_f1scores))

AVG: Scores = 0.9266666666666667 F1-Scores = 0.9246549746549746
AVG: Scores = 0.9466666666666667 F1-Scores = 0.9466006216006215
AVG: Scores = 0.9200000000000002 F1-Scores = 0.9156399156399158
AVG: Scores = 0.9466666666666667 F1-Scores = 0.9457070707070707
AVG: Scores = 0.9466666666666667 F1-Scores = 0.9439310689310689
AVG: Scores = 0.9333333333333333 F1-Scores = 0.932121212121212
AVG: Scores = 0.9199999999999999 F1-Scores = 0.9188228438228438
AVG: Scores = 0.96 F1-Scores = 0.9597979797979799
AVG: Scores = 0.96 F1-Scores = 0.9595959595959596
AVG: Scores = 0.9200000000000002 F1-Scores = 0.918957523957524
AVG: Scores = 0.9533333333333334 F1-Scores = 0.9524410774410773
AVG: Scores = 0.9133333333333333 F1-Scores = 0.9076479076479076
Wall time: 5h 2min 51s


In [6]:
print_metrics(scores, f1scores)

Scores: [0.9266666666666667, 0.9466666666666667, 0.9200000000000002, 0.9466666666666667, 0.9466666666666667, 0.9333333333333333, 0.9199999999999999, 0.96, 0.96, 0.9200000000000002, 0.9533333333333334, 0.9133333333333333]
Best score: 0.96
F1-Scores: [0.9246549746549746, 0.9466006216006215, 0.9156399156399158, 0.9457070707070707, 0.9439310689310689, 0.932121212121212, 0.9188228438228438, 0.9597979797979799, 0.9595959595959596, 0.918957523957524, 0.9524410774410773, 0.9076479076479076]
Max F1-Score: 0.9597979797979799
Avg score: 0.9372222222222223
Avg F1-Score: 0.9354931796598464


### Different Sigma Q
On [ICQ-Training-Estimator-New-Research Notebook](ICQ-Training-Estimator-New-Research.ipynb), the first research we do is about varying the Sigma_Q params - having different values apart from [1, 1, 1] (which is the original one), and the best result we have is [12, 3, 3] with highest accuracy as 0.933 - therefore we don't expect to have an improvement right now, but we still do the test.

#### Integer params

In [7]:
%%time
scores = []
f1scores = []
for i_random_state in range(10, 70, 5):
    curr_scores, curr_f1scores = executeIrisOneVsRest(random_seed=i_random_state,
                                                classifier_function=execute_classifier_split_input_weight_normal_sigma_q,
                                                sigma_q_weights=[12, 3, 3],
                                                max_iter=2000,
                                                print_each_fold_metric=False,
                                                print_avg_metric=True)
    scores.append(np.mean(curr_scores))
    f1scores.append(np.mean(curr_f1scores))

AVG: Scores = 0.7533333333333333 F1-Scores = 0.7158781033781033
AVG: Scores = 0.7733333333333334 F1-Scores = 0.7499847374847375
AVG: Scores = 0.7866666666666667 F1-Scores = 0.7517824767824768
AVG: Scores = 0.7933333333333332 F1-Scores = 0.7624255374255374
AVG: Scores = 0.7733333333333333 F1-Scores = 0.7382173382173383
AVG: Scores = 0.7933333333333333 F1-Scores = 0.7647556147556147
AVG: Scores = 0.7933333333333332 F1-Scores = 0.7624847374847376
AVG: Scores = 0.7866666666666667 F1-Scores = 0.7543956043956044
AVG: Scores = 0.7866666666666666 F1-Scores = 0.7615588115588116
AVG: Scores = 0.7933333333333333 F1-Scores = 0.7611740111740112
AVG: Scores = 0.7933333333333332 F1-Scores = 0.7688700188700188
AVG: Scores = 0.78 F1-Scores = 0.7524438524438525
Wall time: 4h 22min 9s


In [8]:
print_metrics(scores, f1scores)

Scores: [0.7533333333333333, 0.7733333333333334, 0.7866666666666667, 0.7933333333333332, 0.7733333333333333, 0.7933333333333333, 0.7933333333333332, 0.7866666666666667, 0.7866666666666666, 0.7933333333333333, 0.7933333333333332, 0.78]
Best score: 0.7933333333333333
F1-Scores: [0.7158781033781033, 0.7499847374847375, 0.7517824767824768, 0.7624255374255374, 0.7382173382173383, 0.7647556147556147, 0.7624847374847376, 0.7543956043956044, 0.7615588115588116, 0.7611740111740112, 0.7688700188700188, 0.7524438524438525]
Max F1-Score: 0.7688700188700188
Avg score: 0.7838888888888889
Avg F1-Score: 0.7536642369975702


#### Float params

For the float params, we have the best score to be 0.95833, so it could be that we have an improvement here. This happens for the param [0.7, 0.4, 0]

In [9]:
%%time
scores = []
f1scores = []
for i_random_state in range(10, 70, 5):
    curr_scores, curr_f1scores = executeIrisOneVsRest(random_seed=i_random_state,
                                                classifier_function=execute_classifier_split_input_weight_normal_sigma_q,
                                                sigma_q_weights=[0.7, 0.4, 0.0],
                                                max_iter=2000,
                                                print_each_fold_metric=False,
                                                print_avg_metric=True)
    scores.append(np.mean(curr_scores))
    f1scores.append(np.mean(curr_f1scores))

AVG: Scores = 0.9533333333333334 F1-Scores = 0.9529137529137529
AVG: Scores = 0.9533333333333335 F1-Scores = 0.952996632996633
AVG: Scores = 0.9533333333333334 F1-Scores = 0.9524410774410773
AVG: Scores = 0.9333333333333333 F1-Scores = 0.9323063973063972
AVG: Scores = 0.9400000000000001 F1-Scores = 0.9388215488215488
AVG: Scores = 0.9600000000000002 F1-Scores = 0.9597306397306398
AVG: Scores = 0.9333333333333333 F1-Scores = 0.9317340067340065
AVG: Scores = 0.9533333333333334 F1-Scores = 0.9519360269360269
AVG: Scores = 0.9466666666666667 F1-Scores = 0.9448569948569949
AVG: Scores = 0.9266666666666667 F1-Scores = 0.9253198653198653
AVG: Scores = 0.9466666666666667 F1-Scores = 0.9457070707070706
AVG: Scores = 0.9600000000000002 F1-Scores = 0.9588215488215488
Wall time: 3h 49min 30s


In [10]:
print_metrics(scores, f1scores)

Scores: [0.9533333333333334, 0.9533333333333335, 0.9533333333333334, 0.9333333333333333, 0.9400000000000001, 0.9600000000000002, 0.9333333333333333, 0.9533333333333334, 0.9466666666666667, 0.9266666666666667, 0.9466666666666667, 0.9600000000000002]
Best score: 0.9600000000000002
F1-Scores: [0.9529137529137529, 0.952996632996633, 0.9524410774410773, 0.9323063973063972, 0.9388215488215488, 0.9597306397306398, 0.9317340067340065, 0.9519360269360269, 0.9448569948569949, 0.9253198653198653, 0.9457070707070706, 0.9588215488215488]
Max F1-Score: 0.9597306397306398
Avg score: 0.9466666666666668
Avg F1-Score: 0.9456321302154636


### Search results for Learning Rate

#### Small subset of learning rates
The very first result we get is learning_rate = 0.002, which is different from the article's best 0.01. Let's see how our model deals with that.

In [11]:
%%time
scores = []
f1scores = []
for i_random_state in range(10, 70, 5):
    curr_scores, curr_f1scores = executeIrisOneVsRest(random_seed=i_random_state,
                                                classifier_function=execute_classifier_split_input_weight_normal_sigma_q,
                                                sigma_q_weights=[1, 1, 1],
                                                max_iter=2000,
                                                print_each_fold_metric=False,
                                                print_avg_metric=True,
                                                learning_rate=0.002)
    scores.append(np.mean(curr_scores))
    f1scores.append(np.mean(curr_f1scores))

AVG: Scores = 0.9200000000000002 F1-Scores = 0.9188228438228437
AVG: Scores = 0.9533333333333334 F1-Scores = 0.9524410774410773
AVG: Scores = 0.9533333333333334 F1-Scores = 0.9525757575757575
AVG: Scores = 0.9400000000000001 F1-Scores = 0.9387878787878788
AVG: Scores = 0.9333333333333333 F1-Scores = 0.9319528619528619
AVG: Scores = 0.9466666666666667 F1-Scores = 0.9463299663299664
AVG: Scores = 0.9466666666666667 F1-Scores = 0.9462121212121211
AVG: Scores = 0.9 F1-Scores = 0.8918651718651718
AVG: Scores = 0.9533333333333334 F1-Scores = 0.9521380471380472
AVG: Scores = 0.9666666666666668 F1-Scores = 0.9665319865319866
AVG: Scores = 0.9266666666666667 F1-Scores = 0.9212506012506013
AVG: Scores = 0.9600000000000002 F1-Scores = 0.9597979797979799
Wall time: 3h 58min 14s


In [12]:
print_metrics(scores, f1scores)

Scores: [0.9200000000000002, 0.9533333333333334, 0.9533333333333334, 0.9400000000000001, 0.9333333333333333, 0.9466666666666667, 0.9466666666666667, 0.9, 0.9533333333333334, 0.9666666666666668, 0.9266666666666667, 0.9600000000000002]
Best score: 0.9666666666666668
F1-Scores: [0.9188228438228437, 0.9524410774410773, 0.9525757575757575, 0.9387878787878788, 0.9319528619528619, 0.9463299663299664, 0.9462121212121211, 0.8918651718651718, 0.9521380471380472, 0.9665319865319866, 0.9212506012506013, 0.9597979797979799]
Max F1-Score: 0.9665319865319866
Avg score: 0.9416666666666668
Avg F1-Score: 0.9398921911421912


It seems that we managed to improve a bit more our model, as our first result was avg score 0.9266 and now we got 0.9361 - almost 1% more. Let's keep researching!

#### Biggest subset of learning rates
After our big research on [ICQ-Training-Estimator-New-Research Notebook](ICQ-Training-Estimator-New-Research.ipynb), we reached the conclusion that the learning rate of 0.002 still is the best one. So lets keep going on.

#### Best learning rate and best sigmaQ param

Now let's try to find the best learning rate for our best sigmaQ, which was [0.7, 0.4, 0.0]. Our research indicates that this learning rate is 0.01, which is exactly the default one (already researched). 

What happens then when we try to find the best SigmaQ for the best learning rate (0.002)? Our research indicates that the best params are rather [0.0, 0.7, 0.9] instead of the original [0.7, 0.4, 0.0]. Let's see how it goes!

In [13]:
%%time
scores = []
f1scores = []
for i_random_state in range(10, 70, 5):
    curr_scores, curr_f1scores = executeIrisOneVsRest(random_seed=i_random_state,
                                                classifier_function=execute_classifier_split_input_weight_normal_sigma_q,
                                                sigma_q_weights=[0.0, 0.7, 0.9],
                                                max_iter=2000,
                                                print_each_fold_metric=False,
                                                print_avg_metric=True,
                                                learning_rate=0.002)
    scores.append(np.mean(curr_scores))
    f1scores.append(np.mean(curr_f1scores))

AVG: Scores = 0.9466666666666667 F1-Scores = 0.9448653198653199
AVG: Scores = 0.8866666666666667 F1-Scores = 0.8830887630887633
AVG: Scores = 0.9333333333333333 F1-Scores = 0.9318181818181819
AVG: Scores = 0.9200000000000002 F1-Scores = 0.9195959595959596
AVG: Scores = 0.8600000000000001 F1-Scores = 0.8570875420875421
AVG: Scores = 0.9600000000000002 F1-Scores = 0.9597306397306398
AVG: Scores = 0.9533333333333334 F1-Scores = 0.9533670033670033
AVG: Scores = 0.96 F1-Scores = 0.9587373737373737
AVG: Scores = 0.96 F1-Scores = 0.9597306397306398
AVG: Scores = 0.9733333333333334 F1-Scores = 0.9731313131313131
AVG: Scores = 0.8866666666666665 F1-Scores = 0.8810186110186109
AVG: Scores = 0.9666666666666668 F1-Scores = 0.9659764309764309
Wall time: 4h 2min 54s


In [14]:
print_metrics(scores, f1scores)

Scores: [0.9466666666666667, 0.8866666666666667, 0.9333333333333333, 0.9200000000000002, 0.8600000000000001, 0.9600000000000002, 0.9533333333333334, 0.96, 0.96, 0.9733333333333334, 0.8866666666666665, 0.9666666666666668]
Best score: 0.9733333333333334
F1-Scores: [0.9448653198653199, 0.8830887630887633, 0.9318181818181819, 0.9195959595959596, 0.8570875420875421, 0.9597306397306398, 0.9533670033670033, 0.9587373737373737, 0.9597306397306398, 0.9731313131313131, 0.8810186110186109, 0.9659764309764309]
Max F1-Score: 0.9731313131313131
Avg score: 0.9338888888888888
Avg F1-Score: 0.9323456481789815


### Conclusion

Our current best result is:
- Sigma Q params = [0.7, 0.4, 0]
- Learning rate = 0.01
- \# of training epochs = 2000

## Polar coordinates SigmaQ
Let's see now how does our classifier behaves with sigmaQ = rx\*sigmaX + ry\*sigmaY + rz\*sigmaZ

### Importing method

In [6]:
from helpers.icq_executions import execute_classifier_split_input_weight_polar_sigma_q

### Initial execution (sigmaQ = [1, pi/4, pi/4])

In [17]:
%%time
scores = []
f1scores = []
for i_random_state in range(10, 70, 5):
    curr_scores, curr_f1scores = executeIrisOneVsRest(random_seed=i_random_state,
                                                classifier_function=execute_classifier_split_input_weight_polar_sigma_q,
                                                sigma_q_weights=[1, np.pi/4, np.pi/4],
                                                max_iter=2000,
                                                print_each_fold_metric=False,
                                                print_avg_metric=True)
    scores.append(np.mean(curr_scores))
    f1scores.append(np.mean(curr_f1scores))

AVG: Scores = 0.8866666666666667 F1-Scores = 0.8843110593110592
AVG: Scores = 0.9333333333333332 F1-Scores = 0.9313973063973062
AVG: Scores = 0.9200000000000002 F1-Scores = 0.9184848484848486
AVG: Scores = 0.9400000000000001 F1-Scores = 0.9397138047138046
AVG: Scores = 0.9400000000000001 F1-Scores = 0.9388215488215488
AVG: Scores = 0.9600000000000002 F1-Scores = 0.9597306397306398
AVG: Scores = 0.9533333333333334 F1-Scores = 0.9524410774410773
AVG: Scores = 0.9400000000000001 F1-Scores = 0.9378199578199578
AVG: Scores = 0.9466666666666667 F1-Scores = 0.9462121212121213
AVG: Scores = 0.8999999999999998 F1-Scores = 0.8991919191919191
AVG: Scores = 0.9133333333333334 F1-Scores = 0.9114983164983164
AVG: Scores = 0.9533333333333334 F1-Scores = 0.9520875420875422
Wall time: 4h 50min 4s


In [18]:
print_metrics(scores, f1scores)

Scores: [0.8866666666666667, 0.9333333333333332, 0.9200000000000002, 0.9400000000000001, 0.9400000000000001, 0.9600000000000002, 0.9533333333333334, 0.9400000000000001, 0.9466666666666667, 0.8999999999999998, 0.9133333333333334, 0.9533333333333334]
Best score: 0.9600000000000002
F1-Scores: [0.8843110593110592, 0.9313973063973062, 0.9184848484848486, 0.9397138047138046, 0.9388215488215488, 0.9597306397306398, 0.9524410774410773, 0.9378199578199578, 0.9462121212121213, 0.8991919191919191, 0.9114983164983164, 0.9520875420875422]
Max F1-Score: 0.9597306397306398
Avg score: 0.9322222222222224
Avg F1-Score: 0.9309758451425117


### Research result

In [7]:
%%time
scores = []
f1scores = []
for i_random_state in range(10, 70, 5):
    curr_scores, curr_f1scores = executeIrisOneVsRest(random_seed=i_random_state,
                                                classifier_function=execute_classifier_split_input_weight_polar_sigma_q,
                                                sigma_q_weights=[1, 1.2, 2.8],
                                                max_iter=2000,
                                                print_each_fold_metric=False,
                                                print_avg_metric=True,
                                                learning_rate=0.04)
    scores.append(np.mean(curr_scores))
    f1scores.append(np.mean(curr_f1scores))

AVG: Scores = 0.9533333333333334 F1-Scores = 0.9524410774410775
AVG: Scores = 0.9600000000000002 F1-Scores = 0.9597306397306398
AVG: Scores = 0.9400000000000001 F1-Scores = 0.9395286195286194
AVG: Scores = 0.9266666666666667 F1-Scores = 0.9250841750841751
AVG: Scores = 0.9400000000000001 F1-Scores = 0.9385521885521886
AVG: Scores = 0.9466666666666667 F1-Scores = 0.9462626262626262
AVG: Scores = 0.9266666666666665 F1-Scores = 0.9251359751359752
AVG: Scores = 0.9466666666666667 F1-Scores = 0.946127946127946
AVG: Scores = 0.9533333333333334 F1-Scores = 0.9530808080808081
AVG: Scores = 0.9466666666666669 F1-Scores = 0.9458754208754208
AVG: Scores = 0.9400000000000001 F1-Scores = 0.9397138047138048
AVG: Scores = 0.9333333333333333 F1-Scores = 0.931085951085951
Wall time: 4h 40min 41s


In [8]:
print_metrics(scores, f1scores)

Scores: [0.9533333333333334, 0.9600000000000002, 0.9400000000000001, 0.9266666666666667, 0.9400000000000001, 0.9466666666666667, 0.9266666666666665, 0.9466666666666667, 0.9533333333333334, 0.9466666666666669, 0.9400000000000001, 0.9333333333333333]
Best score: 0.9600000000000002
F1-Scores: [0.9524410774410775, 0.9597306397306398, 0.9395286195286194, 0.9250841750841751, 0.9385521885521886, 0.9462626262626262, 0.9251359751359752, 0.946127946127946, 0.9530808080808081, 0.9458754208754208, 0.9397138047138048, 0.931085951085951]
Max F1-Score: 0.9597306397306398
Avg score: 0.9427777777777778
Avg F1-Score: 0.9418849360516027
